src/server/campaignd/server.cpp

```diff
@@ -220,6 +222,73 @@ void server::load_config()
 	// Since an addon is sent in a single WML document this essentially limits
 	// the maximum size of an addon that can be uploaded.
 	simple_wml::document::document_size_limit = cfg_["document_size_limit"].to_int(default_document_size_limit);
+
+	//Loading addons
+	addons_.clear();
+	std::vector<std::string> legacy_addons, dirs;
+	filesystem::get_files_in_dir("data", &legacy_addons, &dirs);
+	config meta;
+	for(const std::string& addon_dir : dirs) {
+		in = filesystem::istream_file(filesystem::normalize_path("data/" + addon_dir + "/addon.cfg"));
+		read(meta, *in);
+		if(meta) {
+			addons_.emplace(meta["name"].str(), meta);
+		} else {
+			WRN_CS << "Failed to load addon from dir '" << addon_dir << "'\n";
+		}
+	}
+
+	// Convert all legacy addons to the new format on load
+	if(cfg_.has_child("campaigns")) {
+		WRN_CS << "Old format addons have been detected in the config! They will be converted to the new file format! ";
+		config& campaigns = cfg_.child("campaigns");
+		WRN_CS << campaigns.child_count("campaign") << " entries to be processed.\n";
+		for(config& campaign : campaigns.child_range("campaign")) {
+			const std::string& addon_id = campaign["name"].str();
+			const std::string& addon_file = campaign["filename"].str();
+			if(get_addon(addon_id)) {
+				ERR_CS << "The addon '" << addon_id
```

Comment:
This is an abnormal state condition. At this point the server needs to die with an error instead of engaging in destructive behaviour since it means someone messed something up in either the code or the filesystem.

Reply:
Added errors in such destructive situations in ba2bbbab21c7871c3e000138803af061648ad92e.